liberum_core/src/swarm_runner/behaviour/kademlia.rs

```diff
@@ -1,87 +1,353 @@
                 }
             }
-            kad::Event::OutboundQueryProgressed {
-                id,
-                result:
-                    kad::QueryResult::GetProviders(Ok(
-                        kad::GetProvidersOk::FinishedWithNoAdditionalRecord { .. },
-                    )),
-                ..
-            } => {
-                debug!("Get providers didn't find any new records");
-                if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
-                    let _ = sender.send(HashSet::new()).inspect_err(|e| {
-                        debug!(
-                            qid = format!("{id}"),
-                            err = format!("{e:?}"),
-                            "Channel closed"
-                        )
-                    });
+        } else {
+            debug!(qid = format!("{id}"), "Channel closed");
+        }
+    }
+
+    fn handle_outbound_query_progressed_put_record(
+        &mut self,
+        id: QueryId,
+        result: Result<PutRecordOk, PutRecordError>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        if result.is_ok() {
+            info!(node = self.node.name, id = format!("{id:?}"), "Put file");
+        } else {
+            error!(
+                node = self.node.name,
+                id = format!("{id:?}"),
+                err = format!("{result:?}"),
+                "Failed to put file"
+            );
+            self.print_neighbours();
+        }
+
+        let sender = self.behaviour.pending_publish_file.remove(&id);
+
+        if let Some(sender) = sender {
+            match result {
+                Ok(_) => {
+                    let _ = sender.send(Ok(()));
+                }
+                Err(e) => {
+                    let _ = sender.send(Err(anyhow!(e)));
                 }
             }
-            kad::Event::InboundRequest {
-                request: kad::InboundRequest::GetProvider { .. },
-            } => {
-                debug!(node = self.node.name, "Received GetProvider")
+        } else {
+            debug!(
+                node = self.node.name,
+                qid = format!("{id}"),
+                "Put Record Progressed: Channel closed"
+            );
+        }
+    }
+
+    fn handle_outbound_query_progressed_get_providers_found(
+        &mut self,
+        id: QueryId,
+        _key: RecordKey,
+        providers: HashSet<PeerId>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
+            let _ = sender.send(providers).inspect_err(|e| {
+                debug!(
+                    node = self.node.name,
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+            // Finish the query to prevent from triggering FinishedWithNoAdditionalRecord
```

Comment:
I don't fully understand how this query progresses yet. It doesn't seem to be well documented. The example usage from libp2p-rust repo finishes the query here. The events structure suggests that `QueryResult::GetProviders` means the query was finished as it contains a `GetProvidersResult` which is a Result, of either `GetProvidersOk::FinishedWithNoAdditionalRecord` or `GetProvidersOk::FoundProviders`.

I think when writing the comment I understood that the query would continue as long as it finds any new provider and stop with `FinishedWithNoAdditionalRecord` when no more is found. The finish there is called because the `FinishedWithNoAdditionalRecord` is also handled and it too sends a response to the node calling `GetProviders`. That's the way they demonstrated the usage of GetProviders in libp2p-rust file-sharing example.

How to best handle this request is not apparent to me and I haven't found anything useful in the documentation. Maybe analysing more examples would help. Either way it's working and fixing it now would probably take too much time. Let's add an issue for that ok?

Reply:
https://github.com/Liberum-Neto/liberum-neto/issues/49